runtime/loader-binary/loader.S

```diff
@@ -0,0 +1,336 @@
+#include "csr.h"
+
+#define SBI_EXT_EXPERIMENTAL_KEYSTONE_ENCLAVE 0x08424b45
+#define SBI_SM_EXIT_ENCLAVE      3006
+# define RUNTIME_VA_START 0xffffffffc0000000 
+
+#if __riscv_xlen == 64
+#define STORE sd
+#define LOAD ld
+#define LOG_REGBYTES  3
+#define WORD .dword
+#elif __riscv_xlen == 32
+#define STORE sw
+#define LOAD lw
+#define LOG_REGBYTES  2
+#define WORD .word
+#endif
+
+#define REGBYTES (1<<LOG_REGBYTES)
+
+.section .text
+_start:
+  // a1: dram_base
+  // a2: dram_size
+  // a3: runtime_base
+  // a4: user_base
+  // a5: free_base
+  // a6: untrusted_ptr
+  // a7: untrusted_size
+  // ...
+
+  // sp <- stack_addr + dram_base (a1)
+  la sp, _estack
+  # add sp, sp, a1
+
+  // save all args to stack 
+  addi sp, sp, -(REGBYTES*7)
+  STORE a1, 0(sp)
+  STORE a2, 1*REGBYTES(sp)
+  STORE a3, 2*REGBYTES(sp)
+  STORE a4, 3*REGBYTES(sp)
+  STORE a5, 4*REGBYTES(sp)
+  STORE a6, 5*REGBYTES(sp)
+  STORE a7, 6*REGBYTES(sp)
+
+  // csrr t0, sepc // user_entry previously written to sepc
+  // addi sp, sp, -(REGBYTES) 
+  // STORE t0, (sp)
+
+  // call load_runtime
+  call load_runtime 
+
+  // exit if errors
+  blt a0, x0, exit
+
+  // switch to va and jump to runtime code 
+  li t0, RUNTIME_VA_START
+  csrw stvec, t0  // store runtime start addresss in stvec 
+
+  la a0, root_page_table  
+  call satp_new // construct new satp
+
+  // set arguments for eyrie_boot
+  LOAD a1, 0(sp)
+  LOAD a2, 1*REGBYTES(sp)
+  LOAD a3, 2*REGBYTES(sp)
+  LOAD a4, 3*REGBYTES(sp)
+  LOAD a5, freeList
+  LOAD a6, 5*REGBYTES(sp)
+  LOAD a7, 6*REGBYTES(sp)
+
+  sfence.vma
+  csrw satp, a0 // switch to virtual addresssing 
+  sfence.vma
+
+exit:
+  call error_and_loop
+  
+_sstack:
```

Comment:
Hmm, ok, I believe this is an old trick that @cathylu10 used.
If I remember correctly, the linker script (.lds file) does specify the stack region, but it didn't work correctly.
For some reason, the stack space is not allocated when the host copies the loader, and that's why we added this hack.

I can't remember what the plan was back then, but we should definitely remove this hack.
I'll talk to you offline.